lib/puppet/parser/scope.rb

```diff
@@ -587,7 +587,7 @@ def define_settings(type, params)
     }
   end
 
-  RESERVED_VARIABLE_NAMES = ['trusted', 'facts'].freeze
+  RESERVED_VARIABLE_NAMES = ['trusted', 'facts', 'server_facts'].freeze
```

Comment:
This is not correct since 'trusted' and 'facts' are always reserved and 'server_facts' are optional.
This means that the test (where RESERVED_VARIABLE_NAMES) is used must also have a conditional check that it is not 'server_facts'. Once this is no longer optional (in puppet 5.0.0), it can be done as above (but not now).